### PR TITLE
[BugFix] Add on_epoch_end hook after on_test/validation_epoch_end hook

### LINKED ISSUE #5917: Unexpected behaviour of hooking inside callback
## 🐛 Bug

`on_epoch_end` hook in callback called between train and validation epochs.

### To Reproduce

Use following callback
```(python)
class BuggyCallback(Callback):
    def on_epoch_end(self, trainer: Trainer, pl_module: LightningModule):
        print("I'm called")
```

### Expected behavior

Line with `I'm called` should appear after train and validation cycles. Instead of this, that line printed exactly after the training cycle, before validation.

### Environment

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): MacOS BigSur
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): -
 - Python version: 3.8.5
 - CUDA/cuDNN version: CPU version
 - GPU models and configuration: -
 - Any other relevant information: -

### Additional context

I checked manually that it works properly in the 1.1.7 version. Thus, the bug appears only in the last version of lightning.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR add missing on_epoch_end hook after validation / test epoch.

Fixes #5917

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by kaushikb11 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Review by awaelchli — COMMENTED** [review]
~it looks like there is on_test_epoch_end but no on_train_epoch_end~
edit: no it's there 😄
